### PR TITLE
Add `NativeArrayExtension`

### PR DESCRIPTION
**Pull request by mbaker3**
Add `NativeArrayExtension`. A collection of extension methods to help developers work with `NativeArray<T>`s

#### Tag Alongs
 - Updated `RandomExtension` to match anvil extension class convention.
 - `CopyFromSingletonBufferJob` - Fixed spelling mistake in docs.

### What is the current behaviour?
There are no existing extension methods provided by Anvil.

### What is the new behaviour?
Developers now have access to `FloodClear()` and `FloodSet()` methods to clear and set the value of a range of elements in a `NativeArray<T>` efficiently.

Some quick tests show that this implementation is ~8x faster than using a `MemsetNativeArray` job.
```
Rough metrics on an array of 262,144 byte elements (ticks)
 - MemsetNativeArray.Schedule: 868
 - FloodSet: 103
 - FloodClear: 72
```
If needed, in the future, we could create a job that executes these more efficient methods so that multiple clears/sets could be done in parallel but we'll build that feature if/when the need comes up. You'd have to be setting values on many truly massive arrays at once to see benefits.

**Discussion:** Open to method naming suggestions. `Flood*` works but I'm not 💯  on it.

### What issues does this resolve?
Convenience and performance in flood filling `NativeArray<T>`s

### What PRs does this depend on?
 - None

### Does this introduce a breaking change?
 - [ ] Yes 
 - [x] No

### CODE REVIEW

File: Scripts/Runtime/Collections/NativeArrayExtension.cs [hunk 1]
```diff
@@ -0,0 +1,62 @@
+using UnityEngine;
+using Unity.Collections;
+using Unity.Collections.LowLevel.Unsafe;
+
+
+/// <summary>
+/// Extension methods for use with a <see cref="NativeArray{T}"/> instance.
+/// </summary>
+public static class NativeArrayExtension
```
Comment (jkeon) [inline Scripts/Runtime/Collections/NativeArrayExtension.cs:9]: Is it `Extension` or `Extensions`?

We seem to be using both? Doesn't matter to me, just clarifying.

File: Scripts/Runtime/Util/Random/RandomExtension.cs [hunk 2]
```diff
@@ -0,0 +1,36 @@
+using System.Runtime.CompilerServices;
+using Unity.Jobs.LowLevel.Unsafe;
+using Unity.Mathematics;
+
+/// <summary>
+/// Extension methods for use with a <see cref="Random"/> instance.
+/// </summary>
+public static class RandomExtension
```
Comment (jkeon) [inline Scripts/Runtime/Util/Random/RandomExtension.cs:8]: Is it `Extension` or `Extensions`?

We seem to be using both? Doesn't matter to me, just clarifying.

**Review by jkeon — COMMENTED** [review]
Looks good, just waiting on the resolution of https://github.com/decline-cookies/anvil-csharp-core/pull/82 to approve.

I see what you mean on Flood but I can't think of anything else good and I think this works.
Makes me think of FloodFill from Photoshop and the analogy works well.

File: Scripts/Runtime/Collections/NativeArrayExtension.cs (hunk 1 above)
Comment (mbaker3) [reply Scripts/Runtime/Collections/NativeArrayExtension.cs:9]: Where are we using Extensions? We aren't as far as I know.
Some 3rd party libraries that we depend on do add the `s`

File: Scripts/Runtime/Util/Random/RandomExtension.cs (hunk 2 above)
Comment (mbaker3) [reply Scripts/Runtime/Util/Random/RandomExtension.cs:8]: See other comment for discussion

File: Scripts/Runtime/Collections/NativeArrayExtension.cs [hunk 3]
```diff
@@ -0,0 +1,62 @@
+using UnityEngine;
+using Unity.Collections;
+using Unity.Collections.LowLevel.Unsafe;
+
+
```
[the remark this answers is not in the document]
Comment (mbaker3) [reply Scripts/Runtime/Collections/NativeArrayExtension.cs:5]: Will move to `namespace Anvil.Unity.DOTS.Collections`

File: Scripts/Runtime/Util/Random/RandomExtension.cs [hunk 4]
```diff
@@ -0,0 +1,36 @@
+using System.Runtime.CompilerServices;
+using Unity.Jobs.LowLevel.Unsafe;
+using Unity.Mathematics;
+
```
[the remark this answers is not in the document]
Comment (mbaker3) [reply Scripts/Runtime/Util/Random/RandomExtension.cs:4]: Will move to `namespace Anvil.Unity.DOTS.Collections`

**Comment by mbaker3** [comment]
Merging in. Will deal with namespace fixes in a separate PR.

**Pull request merged**